### PR TITLE
Use in-memory cache for the auth server API.

### PR DESCRIPTION
**Pull request by klizhentas**
This commit expands the usage of the caching layer
for auth server API:

* Introduces in-memory cache that is used to serve all
Auth server API requests. This is done to achieve scalability
on 10K+ node clusters, where each node fetches certificate authorities,
roles, users and join tokens. It is not possible to scale
DynamoDB backend or other backends on 10K reads per seconds
on a single shard or partition. The solution is to introduce
an in-memory cache of the backend state that is always used
for reads.

* In-memory cache has been expanded to support all resources
required by the auth server.

* Experimental `tctl top` command has been introduced to display
common single node metrics.

Replace SQLite Memory Backend with BTree

SQLite in memory backend was suffering from
high tail latencies under load (up to 8 seconds
in 99.9%-ile on load configurations).

This commit replaces the SQLite memory caching
backend with in-memory BTree backend that
brought down tail latencies to 2 seconds (99.9%-ile)
and brought overall performance improvement.

### CODE REVIEW

**Comment by klizhentas** [comment]
retest this please

File: lib/auth/proto/auth.proto [hunk 1]
```diff
@@ -29,13 +32,45 @@ message Event {
...
+// WatchKind specifies resource kind to watch
+message WatchKind {
+    // Kind is a resource kind to watch
+    string Kind = 1 [(gogoproto.jsontag) = "kind"];
+    // LoadSecrets specifies whether to load secrets
+    bool LoadSecrets = 2 [(gogoproto.jsontag) = "load_secrets"];
+	// Name is an optional specific resource type to watch,
+	// if specified only the events with a specific resource
+	// name will be sent
+	string Name = 3  [(gogoproto.jsontag) = "name"];
```
Comment (klizhentas) [inline lib/auth/proto/auth.proto]: fix formatting

**Review by r0mant — APPROVED** [review]

**Review by russjones — APPROVED** [review]

File: lib/auth/init.go [hunk 2]
```diff
@@ -639,6 +602,34 @@ func (i *Identity) String() string {
...
 }
 
+// CertInfo returns diagnostic information about certificate
+func CertInfo(cert *x509.Certificate) string {
+	return fmt.Sprintf("cert(%v issued by %v:%v)", cert.Subject.CommonName, cert.Issuer.CommonName, cert.Issuer.SerialNumber)
+}
+
+// TLSCertInfo returns diagnostic information about certificate
+func TLSCertInfo(cert *tls.Certificate) string {
+	x509cert, err := x509.ParseCertificate(cert.Certificate[0])
```
[the remark this answers is not in the document]
Comment (klizhentas) [reply lib/auth/init.go]: very rare chance, as tls.Certificate should at least have one cert in the chain

**Comment by klizhentas** [comment]
retest this please

**Comment by klizhentas** [comment]
retest this please

**Pull request merged**